miner/block_builder.go

```diff
@@ -0,0 +1,192 @@
+package miner
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/davecgh/go-xdr/xdr2"
+	"github.com/spacemeshos/go-spacemesh/common"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/mesh"
+	"github.com/spacemeshos/go-spacemesh/p2p"
+	"github.com/spacemeshos/go-spacemesh/state"
+	meshSync "github.com/spacemeshos/go-spacemesh/sync"
+	"math/big"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+const MaxTransactionsPerBlock = 200 //todo: move to config
+
+type BlockBuilder struct{
+	beginRoundEvent chan mesh.LayerID
+	stopChan		chan struct{}
+	newTrans		chan *state.Transaction
+	hareResult		chan HareResult
+	transactionQueue []SerializableTransaction
+	hareRes			HareResult
+	mu sync.Mutex
+	network p2p.Service
+	weakCoinToss	WeakCoinProvider
+	orphans			OrphanBlockProvider
+	started			bool
+}
+
+type Block struct {
+	Id         	mesh.BlockID
+	LayerIndex 	mesh.LayerID
+	Data       	[]byte
+	Coin       	bool
+	Timestamp  	time.Time
+	Txs			[]SerializableTransaction
+	HareResult	[]mesh.BlockID
+	Orphans		[]mesh.BlockID
+}
+
+type SerializableTransaction struct{
+	AccountNonce 	uint64
+	Price			[]byte
+	GasLimit		uint64
+	Recipient 		*common.Address
+	Origin			common.Address //todo: remove this, should be calculated from sig.
+	Amount       	[]byte
+	Payload      	[]byte
+}
+
+
+func NewBlockBuilder(net p2p.Service, beginRoundEvent chan mesh.LayerID, weakCoin WeakCoinProvider, orph OrphanBlockProvider) BlockBuilder{
+	return BlockBuilder{
+		beginRoundEvent:beginRoundEvent,
+		stopChan:make(chan struct{}),
+		newTrans:make(chan *state.Transaction),
+		hareResult:make(chan HareResult),
+		transactionQueue:make([]SerializableTransaction,0,10),
+		hareRes:nil,
+		mu:sync.Mutex{},
```

Comment:
you don't have to do this, in go mutex's nil value is a working mutex. when initializing structs without names, you must specify every field. so that makes you must init even nil values but that's not the case here.

Reply:
is there a functional reason not to initialize a value explicitly?